### PR TITLE
Improve tiled reduction fusion scoring

### LINKED ISSUE #148682: can we make inductor create faster fusions for tiled reductions across dim0 and dim1?
### 🐛 Describe the bug

Can we make fusions of reductions across Mx1 and 1xM tiles fast in inductor?  The key use case for this is scaling tensors to MX across both dim0 and dim1 at the same time, which is important for microscaling (MX) training.  Here is an example snippet to demonstrate a simplified version of the pattern:

```python
def scale_dim0_dim1_reference(x_hp: torch.Tensor, block_size) -> Tuple[torch.Tensor, torch.Tensor]:

    # normalize across dim0
    x_hp_d0_block = x_hp.reshape(-1, block_size)
    x_hp_d0_block_abs = x_hp_d0_block.abs()
    amax_dim0 = torch.amax(x_hp_d0_block_abs, dim=1).unsqueeze(1)
    x_hp_d0_block_normalized = x_hp_d0_block / amax_dim0
    x_hp_d0_normalized = x_hp_d0_block_normalized.reshape(x_hp.shape)

    # normalize across dim1
    x_hp_d1 = x_hp.t().contiguous()
    x_hp_d1_block = x_hp_d1.reshape(-1, block_size)
    x_hp_d1_block_abs = x_hp_d1_block.abs()
    amax_dim1 = torch.amax(x_hp_d1_block_abs, dim=1).unsqueeze(1)
    x_hp_d1_block_normalized = x_hp_d1_block / amax_dim1
    x_hp_d1_normalized = x_hp_d1_block_normalized.reshape(x_hp_d1.shape)

    return x_hp_d0_normalized, x_hp_d1_normalized.t(), amax_dim0, amax_dim1
```

In the code above, we
* start with a tensor and a block size (32 for MX)
* for dim0, partition the tensor into chunks of block_size, normalize by the max absolute value in each block, and write out a normalized tensor and the scales used for normalization
* for dim1, repeat ^

Note: in the "real" use case for MX (https://github.com/pytorch/ao/issues/1788), the differences are:
1. instead of "calculate max absolute value", we will do "calculate MX e8m0 scale"
2. instead of "write out a normalized tensor", we will do "write out a normalized low precision tensor"
3. we also need to swizzle the scales, but that can be done separately from this issue

When I run `torch.compile` on the above example kernel today, I see two kernels - one for each dim.

Claude and I wrote a triton kernel to load the input data in tiles and do the normalization across dim0 and dim1 inline: https://gist.github.com/vkuzo/a7374b1f1f5eabff4a6d774972248c22 / https://github.com/vkuzo/pytorch_scripts/blob/6c26861f2a7d0d31930006b63e538d56026b8aba/mx_cast_poc/20250305_mx_dim0_dim1_cast.py). It seems to be up to 2x faster than the current torch.compile behavior with tile size 32, and up to 4x if we increase tile size to 128, so a faster kernel is definitely possible.  Fusions with preceding op may change the optimal behavior though.

Logs of current torch.compile behavior on example snippet above: https://gist.github.com/vkuzo/7bfd4e23411f22fc25f94323bcd93794

Output of comparison of torch.compile vs triton kernel across a couple of block sizes:

```bash
(pytorch) [vasiliy@devgpu023.atn1 ~/local/pytorch_scripts/mx_cast_poc (20250305_mx_max_dim0_dim1)]$ python 20250305_mx_dim0_dim1_cast.py --M 8192 -K 4096 --BLOCK_SIZE 32
M 8192 K 4096 BLOCK_SIZE 32
GPU: NVIDIA B200
torch version: 2.7.0a0+gitd518490
triton version: 3.2.0
bf16 vs normalized reference sqnrs: dim0 57.75, dim1 57.75
normalized reference vs normalized triton are bitwise equivalent
time_reference_compile_us 182.9869150943399
time_triton_us 94.03462108262123
speedup 1.9459525969011233
(pytorch) [vasiliy@devgpu023.atn1 ~/local/pytorch_scripts/mx_cast_poc (20250305_mx_max_dim0_dim1)]$ python 20250305_mx_dim0_dim1_cast.py --M 8192 -K 4096 --BLOCK_SIZE 64
M 8192 K 4096 BLOCK_SIZE 64
GPU: NVIDIA B200
torch version: 2.7.0a0+gitd518490
triton version: 3.2.0
bf16 vs normalized reference sqnrs: dim0 56.5, dim1 56.5
normalized reference vs normalized triton are bitwise equivalent
time_reference_compile_us 183.88731220657243
time_triton_us 53.66455661375654
speedup 3.42660638249705
(pytorch) [vasiliy@devgpu023.atn1 ~/local/pytorch_scripts/mx_cast_poc (20250305_mx_max_dim0_dim1)]$ python 20250305_mx_dim0_dim1_cast.py --M 8192 -K 4096 --BLOCK_SIZE 128
M 8192 K 4096 BLOCK_SIZE 128
GPU: NVIDIA B200
torch version: 2.7.0a0+gitd518490
triton version: 3.2.0
bf16 vs normalized reference sqnrs: dim0 56.0, dim1 56.0
normalized reference vs normalized triton are bitwise equivalent
time_reference_compile_us 312.7817773722634
time_triton_us 67.17439390386868
speedup 4.656264972332706
(pytorch) [vasiliy@devgpu023.atn1 ~/local/pytorch_scripts/mx_cast_poc (20250305_mx_max_dim0_dim1)]$ python 20250305_mx_dim0_dim1_cast.py --M 8192 -K 4096 --BLOCK_SIZE 256
M 8192 K 4096 BLOCK_SIZE 256
GPU: NVIDIA B200
torch version: 2.7.0a0+gitd518490
triton version: 3.2.0
bf16 vs normalized reference sqnrs: dim0 56.25, dim1 56.25
normalized reference vs normalized triton are bitwise equivalent
time_reference_compile_us 362.2346390041493
time_triton_us 1091.8661034482752
speedup 0.33175738111125397

```

Can we improve this in inductor?

### Versions

main branch

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184128

Allow same-domain sibling reductions that read the same full input buffer through different index expressions to get a non-zero horizontal fusion score. This lets the MX dim0/dim1 block normalization pattern lower to a single Triton kernel instead of two separate kernels.

Fixes #148682

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

File: test/inductor/test_mix_order_reduction.py [hunk 1]
```diff
@@ -263,6 +263,23 @@ def check_one_split_size(split_size):
         check_one_split_size(8)
         check_one_split_size(16)
 
+    def test_block_reduction_transpose_horizontal_fusion(self):
+        def f(x, block_size: int):
```
Comment (eellison) [inline test/inductor/test_mix_order_reduction.py:267]: Do we have perf on this ? Id expect a single load, and, coalesced accesses on both dims, and 3d tile. what is the bandwidth?

**Review by eellison — COMMENTED** [review]
needs perf numbers, probably more work

File: test/inductor/test_mix_order_reduction.py (hunk 1 above)
Comment (jansel) [reply test/inductor/test_mix_order_reduction.py:267]: My agent says I measured the issue-sized bf16 case on an NVIDIA B200 with M=8192, K=4096. The generated kernel is one persistent reduction kernel, but it still has `num_load=2`; it is not the single-load 3D tiled/coalesced-both-dims kernel you described.

Using `score_fusion_memory_threshold=1e12` as the two-kernel baseline vs this PR:
- B=32: 0.0870 ms -> 0.0798 ms, 1.09x, ~3.42 TB/s logical BW if accounting for the current two input reads (~2.58 TB/s if accounting as ideal one-load traffic)
- B=64: 0.1063 ms -> 0.0819 ms, 1.30x, ~3.30 TB/s two-load logical BW
- B=128: 0.1313 ms -> 0.0715 ms, 1.84x, ~3.77 TB/s two-load logical BW

So this PR is an incremental horizontal-fusion/single-launch improvement. The larger single-load 3D tile would need a separate codegen path.

**Review by eellison — COMMENTED** [review]
id still just defer this till we have the right single-load. there is already an option to try to fuse kernels with same buffer read.

**Comment by jansel** [comment]
My agent says I agree this should be deferred rather than landed as an incremental same-buffer-read fusion heuristic. I am closing this PR; issue #148682 should remain open for the proper single-load 3D tiled codegen path.
